vllm/entrypoints/openai/tool_parsers/qwen3coder_xml_tool_parser.py

```diff
@@ -0,0 +1,1116 @@
+            try:
+                float_param_value: float = float(param_value)
+                return float_param_value if float_param_value - int(
+                    float_param_value) != 0 else int(float_param_value)
+            except (ValueError, TypeError):
+                logger.warning(
+                    "Parsed value '%s' of parameter '%s' is not a float "
+                    "in tool '%s', degenerating to string.", param_value)
+            return param_value
+        elif param_type in ['boolean', 'bool', 'binary']:
+            param_value = param_value.lower()
+            return param_value == 'true'
+        else:
+            return param_value
+
+    def _convert_for_json_streaming(self, converted_value: Any,
+                                    param_type: str) -> str:
+        """Convert converted_value based on 
+        whether it's empty and if type is string
+        Args:
+            converted_value: Converted value
+            param_type: Parameter type
+
+        Returns:
+            Converted string for streaming output
+        """
+        # Check if value is empty, but exclude numeric 0
+        if converted_value is None or converted_value == '':
+            return ''
+
+        if param_type in ['string', 'str', 'text', 'varchar', 'char', 'enum']:
+            # String type, remove double quotes
+            return json.dumps(converted_value, ensure_ascii=False)[1:-1]
+        else:
+            # Non-string type, return complete JSON string
+            if not isinstance(converted_value, str):
+                return json.dumps(converted_value, ensure_ascii=False)
+            else:
+                return converted_value
+
+    def _reset_xml_parser_after_tool_call(self):
+        """
+        Each tool_call is treated as a separate XML document, 
+        so we need to reset the parser after each tool_call.
+        """
+
+        with suppress(Exception):
+            self.parser.Parse('', True)
+
+        # recreate XML parser
+        self.parser = ParserCreate()
+        self.setup_parser()
+
+        # Reset current tool_call state
+        if self.current_call_id:
+            self.last_completed_call_id = self.current_call_id
+        self.current_call_id = None
+        self.current_function_name = None
+        self.current_function_open = False
+        self.parameters = {}
+        self.current_param_name = None
+        self.current_param_value = ''
+        self.current_param_value_converted = ''
+        self.current_param_is_first = False
+        self.should_emit_end_newline = False
+        self.start_quote_emitted = False
+
+        # Reset preprocessing and deferred parsing state
+        self._pre_inside_parameter = False
+        self._pre_param_buffer = ""
+        self._pre_current_param_name = None
+        self.defer_current_parameter = False
+        self.deferred_param_raw_value = ""
+
+
+@ToolParserManager.register_module("qwen3_coder_xml")
```

Comment:
Hi @Zhikaiiii, just a small question: the existing `qwen3_coder` should also be contributed by the Qwen team’s @ranpox.
From the unit tests, it looks like `qwen3_coder_xml` can completely replace `qwen3_coder`.

So should we deprecate the existing `qwen3_coder` and adopt the newer, more stable `qwen3_coder_xml` instead?

Reply:
yes, we intent to use qwen3_coder_xml replace qwen3_coder,  but for a clear review and some accuracy problems, we did not directly replace it.  After we fix the accuracy problem, and all review comments been resolved, we will rename it